### DISCUSSION TITLE
Binding Tuya TS0043 switch

### DISCUSSION THREAD

**Discussion by sebo-b** [Koenkk/zigbee2mqtt#5899] [Devices]
Hey,

I'm trying to bind TS0043 to a light switch (TS0012). However, it seems that TS0043 exposes genOnOff as input cluster. It is not shown in webui, "manual" binding via mqtt message fails as well.

Does it mean that TS0043 cannot be bound to any switch or is it a bug, or I don't understand it?

Thanks.

**Comment by sebo-b** [comment]
``` {
    "date_code": "",
    "definition": {
      "description": "Wireless switch with 3 buttons",
      "exposes": [
        {
          "access": 1,
          "description": "Remaining battery in %",
          "name": "battery",
          "property": "battery",
          "type": "numeric",
          "unit": "%",
          "value_max": 100,
          "value_min": 0
        },
        {
          "access": 1,
          "description": "Triggered action (e.g. a button click)",
          "name": "action",
          "property": "action",
          "type": "enum",
          "values": [
            "1_single",
            "1_double",
            "1_hold",
            "2_single",
            "2_double",
            "2_hold",
            "3_single",
            "3_double",
            "3_hold"
          ]
        },
        {
          "access": 1,
          "description": "Link quality (signal strength)",
          "name": "linkquality",
          "property": "linkquality",
          "type": "numeric",
          "unit": "lqi",
          "value_max": 255,
          "value_min": 0
        }
      ],
      "model": "TS0043",
      "vendor": "TuYa"
    },
    "endpoints": {
      "1": {
        "bindings": [],
        "clusters": {
          "input": [
            "genBasic",
            "genPowerCfg",
            "genOnOff"
          ],
          "output": [
            "genOta",
            "genTime"
          ]
        },
        "configured_reportings": []
      },
      "2": {
        "bindings": [],
        "clusters": {
          "input": [
            "genPowerCfg",
            "genOnOff"
          ],
          "output": []
        },
        "configured_reportings": []
      },
      "3": {
        "bindings": [],
        "clusters": {
          "input": [
            "genPowerCfg",
            "genOnOff"
          ],
          "output": []
        },
        "configured_reportings": []
      }
    },
    "friendly_name": "SmartSwitch3",
    "ieee_address": "0x*******",
    "interview_completed": true,
    "interviewing": false,
    "model_id": "TS0043",
    "network_address": *****,
    "power_source": "Battery",
    "supported": true,
    "type": "EndDevice"
  },
```

**Comment by Koenkk** [comment]
To bind it has to be in the output clusters, not input. If it's not in the output cluster it doesn't support it.

**Reply by sebo-b** [reply]
Thanks for the answer.

It is surprising that a switch (battery-powered remote touch button) has genOnOff as input clusters, what input it can take, it basically makes no sense. Is it possible that it is a bug in converters, or you're sure it is reported correctly?

PS: I was able to bind and unbind using deCONZ software - however it didn't have any effect.
PPS: Are you aware of wall switches (physical with relays) that are input and output at the same time? I would like to use 2 the same looking as "1 gang 2 way"?

**Reply by Koenkk** [reply]
Yes, the input and output clusters are not something defined by Zigbee2MQTT, this is what the device sends.

> PPS: Are you aware of wall switches (physical with relays) that are input and output at the same time? I would like to use 2 the same looking as "1 gang 2 way"?

Maybe one of Ubisys (https://www.zigbee2mqtt.io/information/supported_devices.html#ubisys) (I don't have these devices myself)

**Reply by sebo-b** [reply]
Hey,

Yesterday I started reading Zigbee spec and in the ZDO/Service discovery section I have found this (lines 5272-5275):
```
By convention, in cases where the application profile enumerates input clusters and their response
output clusters with the same cluster identifier, the application profile shall list only the input cluster 
within the Simple Descriptor for the purposes of Service Discovery.
```

I'm new to Zigbee but I understand it as if output clusters matches input clusters, they are not separately listed.

Then I made another test, I have hack-fixed `herdsman` to always report input clusters as output clusters:
```--- endpoint.js.orig	2021-01-26 22:18:10.790067742 +0000
+++ endpoint.js	2021-01-25 22:58:40.727720542 +0000
@@ -52,3 +52,3 @@
         this.inputClusters = inputClusters;
-        this.outputClusters = outputClusters;
+        this.outputClusters = outputClusters.concat(inputClusters);
         this.deviceNetworkAddress = deviceNetworkAddress;
```

After that change, I was able to bind and unbind devices. The button was still not able to turn on the light, but I guess that Tuya remote switch is basically not compatible as its intention is to switch scenes (I don't have (yet) Zigbee sniffer to check what exactly is going on). However, this bind was also visible in deCONZ and I was able to unbind there, the second unbind was reporting an error, so definitely bind table was updated.

I'm waiting for Aquara switches to do further testing.

**Reply by okkies** [reply]
so did this ever get fixed? 
Because i dont see any buttons after adding the device to zigbee2mqtt 

i do see the buttons in mqtt rought when i push them 
MQTT publish: topic 'zigbee2mqtt/Lichtknop trap beneden', payload '{"action":"3_single","battery":100,"linkquality":9,"voltage":3000}'

**Reply by samuelthng** [reply]
> Hey,
> 
> Yesterday I started reading Zigbee spec and in the ZDO/Service discovery section I have found this (lines 5272-5275):
> ```
> By convention, in cases where the application profile enumerates input clusters and their response
> output clusters with the same cluster identifier, the application profile shall list only the input cluster 
> within the Simple Descriptor for the purposes of Service Discovery.
> ```
> 
> I'm new to Zigbee but I understand it as if output clusters matches input clusters, they are not separately listed.
> 
> Then I made another test, I have hack-fixed `herdsman` to always report input clusters as output clusters:
> ```--- endpoint.js.orig	2021-01-26 22:18:10.790067742 +0000
> +++ endpoint.js	2021-01-25 22:58:40.727720542 +0000
> @@ -52,3 +52,3 @@
>          this.inputClusters = inputClusters;
> -        this.outputClusters = outputClusters;
> +        this.outputClusters = outputClusters.concat(inputClusters);
>          this.deviceNetworkAddress = deviceNetworkAddress;
> ```
> 
> After that change, I was able to bind and unbind devices. The button was still not able to turn on the light, but I guess that Tuya remote switch is basically not compatible as its intention is to switch scenes (I don't have (yet) Zigbee sniffer to check what exactly is going on). However, this bind was also visible in deCONZ and I was able to unbind there, the second unbind was reporting an error, so definitely bind table was updated.
> 
> I'm waiting for Aquara switches to do further testing. 

@sebo-b Sorry to dig up a 5 year old thread, did you ever manage to get binding working on Aqara devices by using the input clusters?

Would it make sense to raise a PR to include inputClusters into the output cluster list?

**Comment by induna-crewneck** [comment]
Did you ever manage to bind the TS0043?

**Comment by david-alexandru-v** [comment]
Never did. There is also the open issue https://github.com/Koenkk/zigbee2mqtt/issues/23987. Should we wait for the authors to include support? Or are there alternatives?

**Comment by awhitwam** [comment]
This is odd.  I have a working TS0043 at my holiday home that i visit every year.   This year it's stopped working and the  buttons are no longer exposed in zigbee2mqtt.   It must be an update that has broken this.

**Comment by awhitwam** [comment]
just found out it was a breaking change in v2.  fix here; https://community.home-assistant.io/t/using-the-new-action-events-in-zigbee2mqtt-2-0/821709/12?u=vincentbattez

**Comment by sebo-b** [comment]
Hi,

nope - after Koenkk replied me it is not supported, I stopped digging and made a simple automation in NodeRed.

/S
